### PR TITLE
feat: allow custom NSG rules

### LINKED ISSUE #720: Add Custom NSG rules to template
**Is this a request for help?**:
No 
---

**Is this an ISSUE or FEATURE REQUEST?** (choose one): 
FEATURE 

---

**What version of acs-engine?**:
ALL 
---

**Orchestrator and version (e.g. Kubernetes, DC/OS, Swarm)**
ALL

**What you expected to happen**:
When the template is used for generating the resource/deployment files a master NSG is created with some default securitygroup rules; as it is there is no current way to override or add custom rules in the template - they need to be added manually after generation. This would be a very nice feature to add for people trying to lock down vnet/subnets for security + compliance when running ACS clusters for production.

### PR DESCRIPTION
**Pull request by sylr**
**Reason for Change**:

Currently the NSG rules are hardcoded and reapplied every upgrades.

This allow to replace the default rules.

**Issue Fixed**:

Fixes #720 

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Notes**:

```json
{
  "apiVersion": "vlabs",
  "properties": {
    "networkSecurityRules": [
      {
        "name": "allow-ssh-from-work",
        "properties": {
          "description": "allow-ssh-from-work",
          "protocol": "tcp",
          "sourcePortRange": "*",
          "destinationPortRange": "22",
          "sourceAddressPrefix": "192.168.1.0/24"
        }
      }
    ]
  }
}
```

### CODE REVIEW

**Comment by sylr** [comment]
> /shrug

Are you not convinced by the idea or the implementation ?

**Comment by mboersma** [comment]
@sylr no! Sorry, we finally resurrected our prow bot and I was trying to use the `/shrug` command to test it. Apologies for the noise. 😄

**Comment by sylr** [comment]
Ok :)

**Comment by sylr** [comment]
@mboersma the CI does not seem to have started.

**Comment by CecileRobertMichon** [comment]
FYI https://github.com/Azure/aks-engine/pull/1705 might change "and reapplied every upgrades"

**Comment by sylr** [comment]
Guy, any feedbacks about this ?

**Comment by CecileRobertMichon** [comment]
@sylr can you please take a look at #1705 and see if it changes anything in regards to this PR?

**Comment by sylr** [comment]
@CecileRobertMichon I don't think it does.

File: pkg/api/types.go [hunk 1]
```diff
@@ -91,6 +92,7 @@ type Properties struct {
 	AddonProfiles           map[string]AddonProfile  `json:"addonProfiles,omitempty"`
 	FeatureFlags            *FeatureFlags            `json:"featureFlags,omitempty"`
 	CustomCloudProfile      *CustomCloudProfile      `json:"customCloudProfile,omitempty"`
+	NetworkSecurityRules    []network.SecurityRule   `json:"networkSecurityRules,omitempty"`
```
[the remark this answers is not in the document]
Comment (sylr) [reply pkg/api/types.go]: Ok.

File: pkg/engine/networksecuritygroups.go [hunk 2]
```diff
@@ -10,11 +10,19 @@ import (
...
+	var securityRules []network.SecurityRule
+	var sshRule, kubeTLSRule network.SecurityRule
+
 	armResource := ARMResource{
 		APIVersion: "[variables('apiVersionNetwork')]",
 	}
 
-	sshRule := network.SecurityRule{
+	if len(cs.Properties.NetworkSecurityRules) > 0 {
+		securityRules = cs.Properties.NetworkSecurityRules
```
[the remark this answers is not in the document]
Comment (sylr) [reply pkg/engine/networksecuritygroups.go]: Done

**Comment by sylr** [comment]
@CecileRobertMichon @jackfrancis , 

#1705 has been reverted with #1754 making this more important.

Can we move forward ?

**Comment by sylr** [comment]
Guys ?

**Comment by CecileRobertMichon** [comment]
@sylr sorry for the delay, will take a look again today. Can you please rebase to fix the conflict?

**Comment by jackfrancis** [comment]
O.K., so it seems like we're on the same page that this addition is for NSG rules attached to the master vms only. My feedback at this point relates to how we merge these custom rules with the rules that we deliver by default, to avoid rule collisions.

Should we use the rule name as reserved keys? So if you bring your own rules, you are able to "override" the built-in rules by naming your custom rule with the same name (and ostensibly using it to provide similar ACL functionality, allow ssh, for example).

@CecileRobertMichon @mboersma @devigned @PatrickLang thoughts?

And thanks for your patience @sylr as we bikeshed (and or ignore) this PR to near-death. I do think it's a value-add to enable flexible user-provided NSG rules. Ideally we define a common interface that is used in common by both the "built-in" rules, and any user-provide rules, so that we have more control over ensuring that the set of total NSG rules (user-provided plus built-in) are sane and may work together.

**Comment by itzikstein** [comment]
If is not to late 
Can you add to this feature request the following:
In case “Custom VNET” (vnetSubnetId) is defined on the masterProfile 
Move the NSG rule responsibility to the user to set on the  “Custom VNET” provided 
And skip/not set the default securityRules (sshRule,kubeTLSRule,rdpRule)

**Comment by sylr** [comment]
@itzikstein with this PR, if your config does not provide custom rules, the aks-engines default ones will be provisionned, if you specify some, the aks-engines default ones (ssh-all, https-all) **will not** be provisioned.

**Comment by sylr** [comment]
Can I have some feedback about this PR ?

File: pkg/engine/networksecuritygroups.go [hunk 3]
```diff
@@ -10,11 +10,19 @@ import (
...
+	var sshRule, kubeTLSRule network.SecurityRule
+
 	armResource := ARMResource{
 		APIVersion: "[variables('apiVersionNetwork')]",
 	}
 
-	sshRule := network.SecurityRule{
+	if cs.Properties.MasterProfile != nil && len(cs.Properties.MasterProfile.NetworkSecurityRules) > 0 {
+		securityRules = cs.Properties.MasterProfile.NetworkSecurityRules
+		goto NSG
```
Comment (CecileRobertMichon) [inline pkg/engine/networksecuritygroups.go]: I would prefer avoiding the use of goto here, can we have an `else` statement instead?

**Comment by CecileRobertMichon** [comment]
@sylr would love to see this move forward. Do you think you'll have time to iterate on it or would you rather close it and have someone take over? 

What do you think about @jackfrancis' comment above and the idea of merging defaults with user provided keys? I think it would make sense because then the user doesn't have to worry about providing the rules that are necessary for the cluster to work (eg. API Server inbound), they can explicitly override it if needed but by default the rule would still be there. So for example as a user I might want to override the SSH rule but keep the apiserver one the same. In the current state of the PR I would have to specify all the rules (which involves looking at the code to know what rules usually get created) vs. just specifying what I want to change. 

If you're willing to drive this, let's work together to get this merged before the next major release.

**Comment by sylr** [comment]
I don't think I'll have the time to work on this now.

Anyone feel free to pick up the work where I left it.

However I would prefer that custom rules replace default rules altogether.

**Pull request closed without merging**